### PR TITLE
Insert club members into the leaderboard

### PR DESCRIPTION
**Pull request by FabianvZ**
This change allows the user to insert club members into the leaderboard using the settings
image
These players are then placed into the leaderboard
image

### CODE REVIEW

File: API/ClubLeaderboardAPI.as [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            warn("Request is null");
+            return null;
+        }
+        if(mapUID == ""){
+            warn("No map selected");
+            return null;
+        }
+
+        while (!NadeoServices::IsAuthenticated("NadeoLiveServices")) yield();
+        auto req = NadeoServices::Get("NadeoLiveServices", "https://live-services.trackmania.nadeo.live/api/token/leaderboard/group/Personal_Best/map/" + mapUID + "/club/" + clubID + "/top?length=" + length + "&offset=0");
```
Comment (Banalian) [inline API/ClubLeaderboardAPI.as:26]: This could get broken down to different constant. I may be wrong but there's `NadeoServices::BaseX` functions you can use, then i'd suggest a generate url function that takes the mapUID, clubID and length parameters.

Additionally , I don't think you need the IsAuthenticated call, since this is done at the start of the plugin. If you saw any reference that says otherwise that's fine, I'd be curious to know

File: API/ClubLeaderboardAPI.as [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        dictionary usernames = NadeoServices::GetDisplayNamesAsync(playerIds);
+        for (uint i = 0; i < playerIds.Length; i++)
+        {
+            response[i]["username"] = string(usernames[response[i]["accountId"]]);
+        }
+        return response;
+    }
+
+    MwId GetMainUserId() {
+    auto app = cast<CTrackMania>(GetApp());
```
Comment (Banalian) [inline API/ClubLeaderboardAPI.as:66]: you missed the indentation here. Additionally, where is this used?

File: Controller/Main.as [hunk 3]
```diff
@@ -122,6 +122,31 @@ void RefreshLeaderboard(){
...
+                tmpEntry.positionData = allClubData[i];
+
+                if (processedAccountIds.Find(clubLeaderboard[j]["accountId"]) != -1) {
+                    continue;
+                }
+                req.scores.InsertLast(tmpEntry.time);
+                processedAccountIds.InsertLast(clubLeaderboard[j]["accountId"]);
+                clubEntries.InsertLast(tmpEntry);
+            }
+        }
```
Comment (Banalian) [inline Controller/Main.as:148]: What if this was also part of the ClubLeaderboardAPI ? it would help with this function already being really bulky. You could have all the logic in the Club namespace, then here call `ClubLeaderboardAPI::GetClubLeaderboard` which would return a list of `LeaderboardEntry@` .

I could see this being nice, but I'm open for debate hehe

File: Model/Data.as [hunk 4]
```diff
@@ -153,4 +153,6 @@ bool failedRefresh = false;
 
 EnumCurrentMode currentMode = EnumCurrentMode::INVALID;
 
-bool forceRefreshAfterSurroundFail = false;
\ No newline at end of file
+bool forceRefreshAfterSurroundFail = false;
+
+array<PositionData> allClubData = {};
```
Comment (Banalian) [inline Model/Data.as:158]: This can be moved below line 104

File: Render/RenderSettings.as [hunk 5]
```diff
@@ -383,6 +383,28 @@ void RenderPositionDataCustomization(){
     changed = GetPositionData("S314ke Medal", 10008, s314keMedalPositionData) || changed;
 #endif
 
+    UI::Separator();
```
Comment (Banalian) [inline Render/RenderSettings.as:386]: I'm not sure what's the best place UX-wise, but could you add a setting to choose the max amount of club entry to gather? make sure to have a defined range so people can't put -5 or 500 on it please

File: Render/RenderSettings.as [hunk 6]
```diff
@@ -383,6 +383,28 @@ void RenderPositionDataCustomization(){
...
+    }
+    if(UI::Button("- : Remove a club")){
+        if(allClubData.Length > 0){
+            allClubData.RemoveAt(allClubData.Length - 1);
+            OnSettingsChanged();
+        }
+    }
+
+    for(uint i = 0; i < allClubData.Length; i++){
+        changed == GetPositionData("Club id " + (i+1), -i, allClubData[i], true) || changed;
```
Comment (Banalian) [inline Render/RenderSettings.as:405]: nitpick, could you have the id start at either 10100 or 20k please ? I don't really mind the number, as long as it's positive and above the last medal one

File: exclude.txt
Comment (Banalian) [inline exclude.txt:1]: nice addition, thank you! I've been meaning to do it at some point but never got around to it...

Might be worth adding .idea to the gitignore too if you've got that folder on your workspace

File: test.bat
Comment (Banalian) [inline test.bat:1]: could you rename it to build.bat?

File: exclude.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+build.bat
```
Comment (Banalian) [inline exclude.txt:1]: what's the difference between build and test?

**Review by Banalian — CHANGES_REQUESTED** [review]
Hi! I ended up adding quite a bunch of comments, but it's a solid base!

I'm glad you went with the namespace idea, makes for a clean feature that gets plugged in the plugin code nicely. Thanks a lot for this feature, I'm sure some people will be very glad to see that!

All the comments are things I'd like to see, but if you have a strong opinion (or if that's somehow not doable) I'm open to talk about it, I just went with how I'd generally structure things to stay close to what's already there.

File: API/ClubLeaderboardAPI.as [hunk 8]
```diff
@@ -0,0 +1,77 @@
+namespace ClubLeaderboardAPI 
+{
+    int counterTriesAPI = 0;
+    int maxTriesAPI = 3;
+    bool failedAPI = false;
+
+    Json::Value@ GetClubLeaderboard(uint clubID, string mapUID, int length = 100){
```
[the remark this answers is not in the document]
Comment (FabianvZ) [reply API/ClubLeaderboardAPI.as]: There is now a settings to control the amount of club members it retrieved per club. The default is 15 and it can only be between 0 and 100

File: API/ClubLeaderboardAPI.as [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        if(!useExternalAPI){
```
[the remark this answers is not in the document]
Comment (FabianvZ) [reply API/ClubLeaderboardAPI.as]: You're right. I have removed this setting

File: API/ClubLeaderboardAPI.as (hunk 1 above)
Comment (FabianvZ) [reply API/ClubLeaderboardAPI.as]: I have created a GenerateUrl method that uses NadeoServices::BaseURLLive()

File: API/ClubLeaderboardAPI.as (hunk 2 above)
Comment (FabianvZ) [reply API/ClubLeaderboardAPI.as]: This was used to retrieve the usernames, but was not needed anymore. I have removed it

File: Controller/Main.as (hunk 3 above)
Comment (FabianvZ) [reply Controller/Main.as]: This has been moved to ClubLeaderboardAPI. GetClubLeaderboardMembers now returns a list of LeaderboardEntry@ for all clubs. The json is parsed to leaderboardEntries in ParseJsonToLeaderboardEntry

File: Render/RenderSettings.as (hunk 5 above)
Comment (FabianvZ) [reply Render/RenderSettings.as:390]: The setting is displayed at the bottom of the general customization
image
The range is 0-100

File: Render/RenderSettings.as (hunk 6 above)
Comment (FabianvZ) [reply Render/RenderSettings.as]: The ids now start at 20k

File: exclude.txt (hunk 7 above)
Comment (FabianvZ) [reply exclude.txt:1]: The test script copies all files (except exclude.txt) to the plugins folder into a folder with the name of the script. The way you can have the test version in a folder and the signed version as a .op file. Openplanet in developer mode loads the folder version if it is present. So that is for easy testing.

In discord rivalry ping I also have a build script which zips the files and renames it to a .op file that I can upload to openplanet.

File: test.bat
Comment (FabianvZ) [reply test.bat:1]: For me it is a test script. I also have a build script that I have not included here. What both do is explained in the next comment. If you then still want it named build.bat I'll do that

**Comment by FabianvZ** [comment]
I'm glad you are excited about the feature. I have changed your requested changes. You still need to decide what you want to do the the test.bat script.

File: Controller/Main.as [hunk 10]
```diff
@@ -209,6 +214,17 @@ void RefreshLeaderboard(){
                     alreadyHandled = medalFound[medalIndex];
                 }
             }
+
+            // Update positions for club members
+            for (uint j = 0; j < clubEntries.Length; j++) {
+                if (clubEntries[j].time == resp.positions[i].time) {
+                    resp.positions[i].positionData = clubEntries[j].positionData;
```
Comment (Banalian) [inline Controller/Main.as:221]: were you able to test the edge case where two club members had the same time? I think you might get the same position data in that case.
If I'm right (it's been a long weekend for me x)), you might need something like an array of bool (the length of clubentries) then say that the J entry has been found when you do find it, then skip the entry for future loops

File: Controller/Main.as [hunk 11]
```diff
@@ -122,6 +122,11 @@ void RefreshLeaderboard(){
             return;
         }
 
+        array<LeaderboardEntry@> clubEntries = ClubLeaderboardAPI::GetClubLeaderboardMembers(req.mapId);
```
Comment (Banalian) [inline Controller/Main.as:125]: Thanks for the changes! 
Regarding this, I should have thought about it before, sorry; Do you think it would be valuable to also offer this for the non-api branch? If you think so and if you wish to add it, remember that you will only be able to request the positions of users who are better than you, since the endpoint used only return accurate positions for times below yours (you can check how I handle medals for that)

File: Render/RenderSettings.as [hunk 12]
```diff
@@ -112,6 +113,9 @@ void RenderSettingsCustomization(){
 
     UI::EndDisabled();
 
+    UI::Text("\n\tClub settings");
+    clubMembersToRetrieve = Math::Max(Math::Min(UI::InputInt("Number of club members to retrieve per club", clubMembersToRetrieve), 100), 0);
```
Comment (Banalian) [inline Render/RenderSettings.as:117]: instead of that min/max call, you could call OnSettingsChanged at the end of that setting tab function, then add a small logic to correctly clamp that value between 0 and 100

File: exclude.txt (hunk 7 above)
Comment (Banalian) [reply exclude.txt:1]: Thank you! In that case, I'll gladly take your build.bat script too please, that's what I've been meaning to add at some point

**Review by Banalian — CHANGES_REQUESTED** [review]
Thanks for all the changes! The only comments left that I'd like to have before merging are a unhandled edge-case and a small setting change, the other two are more optional.

File: test.bat
Comment (Banalian) [reply test.bat:1]: as said in the other comment, I'll gladly take both scripts in that case please, the test script will be useful to other contributors, and that build script will be useful to me :D

**Comment by Banalian** [comment]
@FabianvZ Do you have an ETA for the requested changes? I'm waiting for this PR before making the next update to the game. Additionally, under which name do you want to be credited in the changelog :D ?

**Comment by FabianvZ** [comment]
> @FabianvZ Do you have an ETA for the requested changes? I'm waiting for this PR before making the next update to the plugin. Additionally, under which name do you want to be credited in the changelog :D ?

I'll have it finished by tomorrow. You can credit me as FabianvZ

File: Render/RenderSettings.as (hunk 12 above)
Comment (FabianvZ) [reply Render/RenderSettings.as]: Done
